pkg/distributor/ha_tracker_test.go

```diff
@@ -70,6 +101,196 @@ func checkReplicaTimestamp(t *testing.T, duration time.Duration, c *defaultHaTra
+			ElectedAt:      now + 10,
+			ElectedChanges: 3,
+		}
+	}
+
+	{
+		ours, ch := merge(firstReplica(), firstReplicaWithHigherReceivedAt())
+		assert.Equal(t, expectedFirstAndFirstHigherReceivedAtMerge(), ours)
+		assert.Equal(t, expectedFirstAndFirstHigherReceivedAtMerge(), ch)
+	}
+
+	{ // idempotency: (no change after applying same Replica again)
+		ours, ch := merge(expectedFirstAndFirstHigherReceivedAtMerge(), firstReplica())
+		assert.Equal(t, expectedFirstAndFirstHigherReceivedAtMerge(), ours)
+		assert.Equal(t, (*ReplicaDesc)(nil), ch)
+	}
+
+	{ // commutativity: Merge(firstReplicaWithHigherReceivedAt, first) == Merge(first, firstReplicaWithHigherReceivedAt)
+		our, ch := merge(firstReplicaWithHigherReceivedAt(), firstReplica())
+		assert.Equal(t, expectedFirstAndFirstHigherReceivedAtMerge(), our)
+		// change is nil in this case, since the incoming ReplicaDesc has lower receivedAt timestamp
+		assert.Equal(t, (*ReplicaDesc)(nil), ch)
+	}
+
+	{ // associativity: Merge(Merge(first, second), third) == Merge(first, Merge(second, third))
+		ours1, _ := merge(firstReplica(), secondReplica())
+		ours1, _ = merge(ours1, thirdReplica())
+		assert.Equal(t, expectedFirstSecondThirdAtMerge(), ours1)
+
+		ours2, _ := merge(secondReplica(), thirdReplica())
+		ours2, _ = merge(ours2, firstReplica())
+		assert.Equal(t, expectedFirstSecondThirdAtMerge(), ours2)
+	}
+
+}
+
+func TestHaTrackerWithMemberList(t *testing.T) {
+	var config memberlist.KVConfig
+
+	const cluster = "cluster"
+	const replica1 = "r1"
+	const replica2 = "r2"
+
+	flagext.DefaultValues(&config)
+	ctx := context.Background()
+
+	config.TCPTransport = memberlist.TCPTransportConfig{
+		BindAddrs: getLocalhostAddrs(),
+		BindPort:  0,
+	}
+
+	config.Codecs = []codec.Codec{
+		GetReplicaDescCodec(),
+	}
+
+	memberListSvc := memberlist.NewKVInitService(
+		&config,
+		log.NewNopLogger(),
+		&dnsProviderMock{},
+		prometheus.NewPedanticRegistry(),
+	)
+
+	t.Cleanup(func() {
+		assert.NoError(t, services.StopAndAwaitTerminated(ctx, memberListSvc))
+	})
+
+	c, err := newHaTracker(HATrackerConfig{
+		EnableHATracker: true,
+		KVStore: kv.Config{Store: "memberlist", StoreConfig: kv.StoreConfig{
+			MemberlistKV: memberListSvc.GetMemberlistKV,
+		}},
+		UpdateTimeout:          time.Millisecond * 100,
+		UpdateTimeoutJitterMax: 0,
+		FailoverTimeout:        time.Millisecond * 2,
+	}, trackerLimits{maxClusters: 100}, nil, log.NewNopLogger())
+	require.NoError(t, services.StartAndAwaitRunning(ctx, c))
+	require.NoError(t, err)
+	t.Cleanup(func() {
+		assert.NoError(t, services.StopAndAwaitTerminated(ctx, c))
+	})
+
+	now := time.Now()
+
+	// Write the first time.
+	err = c.checkReplica(context.Background(), "user", cluster, replica1, now)
+	assert.NoError(t, err)
+
+	// Throw away a sample from replica2.
+	err = c.checkReplica(context.Background(), "user", cluster, replica2, now)
+	assert.Error(t, err)
+
+	// Wait more than the overwrite timeout.
+	now = now.Add(1100 * time.Millisecond)
```

Comment:
can you put the timeout in a constant and use it here with `+ time.Millisecond*X`? It's not immediately obvious if you refer to update timeout or failover timeout

Reply:
Done [6cfcd63](https://github.com/grafana/mimir/pull/10020/commits/6cfcd6362e5ac3911cc1a279d7d6c4fba246ead4). 

Given that we want to avoid of being flaky, I added 100ms . Let me know if it makes sense